src/core/ipe-setup.c

```diff
@@ -0,0 +1,110 @@
+                                IPE_INPUT_POLICY_DIR);
+                return 0;
+        }
+
+        policy_dir = opendir(IPE_INPUT_POLICY_DIR);
+        if (!policy_dir)
+                return log_error_errno(
+                                errno,
+                                "Failed to open the IPE custom policy directory %s: %m",
+                                IPE_INPUT_POLICY_DIR);
+
+        FOREACH_DIRENT(de, policy_dir, return log_error_errno(errno, "Failed to read %s directory: %m", IPE_INPUT_POLICY_DIR)) {
+                _cleanup_free_ char *policy_name = NULL, *output_path = NULL, *activate_path = NULL;
+                _cleanup_close_ int input = -EBADF, output = -EBADF;
+                const char *suffix;
+
+                if (de->d_type != DT_REG)
+                        continue;
+
+                suffix = endswith(de->d_name, ".p7b");
+                if (!suffix)
+                        continue;
+
+                policy_name = strndup(de->d_name, suffix - de->d_name);
+                if (!policy_name)
+                        return log_oom();
+
+                input = openat(dirfd(policy_dir), de->d_name, O_RDONLY|O_NOFOLLOW|O_CLOEXEC);
+                if (input < 0)
+                        return log_error_errno(
+                                        errno,
+                                        "Failed to open the IPE custom policy file %s/%s: %m",
+                                        IPE_INPUT_POLICY_DIR,
+                                        de->d_name);
+
+                /* If policy is already installed, try to update it */
+                output_path = path_join(IPE_SECFS_POLICIES, policy_name, "update");
+                if (!output_path)
+                        return log_oom();
+
+                output = open(output_path, O_WRONLY|O_CLOEXEC);
+                if (output < 0 && errno == ENOENT)
+                        /* Policy is not installed, install it and activate it */
+                        output = open(IPE_SECFS_NEW_POLICY, O_WRONLY|O_CLOEXEC);
+                if (output < 0)
+                        return log_error_errno(
+                                        errno,
+                                        "Failed to open the IPE policy handle for writing: %m");
+
+                /* The policy is inline signed in binary format, so it has to be copied in one go, otherwise the
+                 * kernel will reject partial inputs with -EBADMSG. */
+                r = copy_bytes(input, output, UINT64_MAX, /* copy_flags= */ 0);
+                if (r < 0)
+                        return log_error_errno(
+                                        r,
+                                        "Failed to copy the IPE policy %s/%s to %s: %m",
+                                        IPE_INPUT_POLICY_DIR,
+                                        de->d_name,
+                                        output_path);
```

Comment:
hmm, dunno, but after writing the policy, wouldn't it conceptually make sense to close the output fd here? i mean, i unerstand the kernel doesn't really care, but for the casual reader i think it would make sense to add `output = safe_close(output);` here to make clear the policy is now loaded.